The vault issuer can now be given a serviceAccountRef instead of relying on static service account tokens

maelvls opened the pull request:
|  🌟 The solution that this PR implements [Proposed solution: cert-manager requests a token using the Token Request API](https://hackmd.io/@maelvls/cert-manager-vault-issuer-broken-kubernetes-1-24#solution-1) |
|-|

**tl;dr:** With the new `serviceAccountRef` field, cert-manager doesn't requests the token on behalf of the pod in order to authenticate with Vault. Thanks to `serviceAccountRef`, the Vault issuer will stop relying on insecure static tokens, and use short-lived tokens instead.

```yaml
apiVersion: cert-manager.io/v1
kind: Issuer
spec:
  vault:
    path: pki/sign/vault-issuer
    server: http://vault.default.svc.cluster.local:8200
    auth:
      kubernetes:
        role: vault-issuer
        mountPath: /v1/auth/kubernetes
        serviceAccount:
          name: vault-issuer     # ✨
```

The expiration duration and the audience aren't configurable by the user:

- The audience is generated by cert-manager to prevent a malicious user from creating a second issuer with the intent of reusing the existing service account. The format is:

  ```text
  "vault://<namespace>/<issuer-name>"   (for an Issuer)
  "vault://<issuer-name>"               (for a ClusterIssuer)
   ```
   
   Note that this extra security isn't mandatory. You can very well configure the role with an empty audience. That said, we recommend that you use the `audience` field, e.g.:

   ```bash
   vault write auth/kubernetes/role/vault-issuer \
       bound_service_account_names=vault-issuer \
       bound_service_account_namespaces=default \
       audience="vault://default/vault-issuer" \
       policies=vault-issuer \
       ttl=1m
   ```

   **Alternative solution considered:** we considered using a hard-coded `vault` audience, but that did not allow to prevent the "Risk 1" (see below table). The UX would have been better, and would correspond to what people expect from an audience (i.e., the audience should be commanded by Vault to cert-manager, not the other way around). But since this is an extra level of security (extra because Vault won't check the audience by default), it should not affect the UX much.

   <table>
    <tr>
        <th>Risk A</th>
        <th>Risk B</th>
    </tr>
    <tr>
        <td><img alt="token-stealing-audience drawio" src="https://user-images.githubusercontent.com/2195781/207842392-a005de3e-7566-4f82-8294-7779daecdc40.png" width=350></td>
        <td><img alt="token-stealing-audience drawio" src="https://user-images.githubusercontent.com/2195781/207843767-cf348767-8570-4cb9-a76e-6ec247656d61.png" width=350></td>
    </tr>
    <tr>
        <td>
            <strong>Remediation:</strong>
            <ul>
                <li><strong>Restrict Issuer in Vault Role:</strong> Include the issuer name and namespace to the JWT and configure Vault to only accept that issuer name and namespace. For example, use <code>vault://namespace-1/issuer-1</code> as the audience.</li>
                <li><strong>URL lock down:</strong> Using RBACs, make sure the Issuer and ClusterIssuer can only be edited by administrators. That should prevent the possibility of an actor using a URL pointing to a malicious server.</li>
            </ul>
        </td>
        <td>
            <strong>Remediation:</strong>
            <ul>
                <li><strong>Audience lock down:</strong> Use a hardcoded audience such as `vault`, or a hardcoded prefix such as `vault-*`.</li>
                <li><strong>Forbid sensitive audiences:</strong> Refuse the cluster's audience like https://kubernetes.default.svc.cluster.local.</li>
                <li><strong>URL lock down:</strong> Using RBACs, make sure the Issuer and ClusterIssuer can only be edited by administrators. That should prevent the possibility of an actor using a URL pointing to a malicious server.</li>
                <li><strong>One SA per external service:</strong> Enforce that each service account is only ever linked to a single external service. E.g., one service account for Vault, one for GCP, one for AWS.</li>
                <li><strong>Least privileges:</strong> Enforce that the service account attached to that JWT doesn't have any privileges.</li>
            </ul>
        </td>
    </tr>
</table>


- The expiration duration for the Kubernetes tokens is hard-coded to 10 minutes (that's the minimum accepted). The expiration duration for the Vault tokens is 1 minute. The expiration duration is set to 10 minute because the Kubernetes token will be immediately discarded after authenticating with Vault. **No token rotation is done; instead, a new Kubernetes token is requested on every call to Vault.** We have discussed this multiple times, and it seems OK as a first iteration.
- **POSSIBLE BREAKING CHANGE:** There is now webhook-side validation for the `vault.auth` field in the Issuer and ClusterIssuer. Previously, the validation was only performed "controller-side". People were able to create a faulty Issuer or ClusterIssuer, and they won't be able anymore. I made sure that the new validation is looser or equal to the existing controller-side validation so that Issuers and ClusterIssuers that were working previously are still working now.

**Work**

- [x] Unit test ("if kubernetes.serviceAccountRef set, request token and exchange it for a vault token"),
- [x] End-to-end test:
    ```sh
    make e2e GINKGO_FOCUS=".*should be ready with a valid serviceAccountRef"`)
    ```
- [x] Documentation on the website (https://github.com/cert-manager/website/pull/1081)
- [x] I manually tested the feature on these environments: kind cluster (Kubernetes 1.21) with an internal Vault running as a deployment.
- [x] Webhook-side validation (eg only one of [secretRef, serviceAccountRef] can be used).
- [x] Controller-side validation.

**If you want to test this feature, here are the steps:**

```sh
 make -j e2e-setup-certmanager
helm upgrade --install vault hashicorp/vault --set server.dev.enabled=true --set server.logLevel=debug --set global.tlsDisable=true --wait
kubectl wait --for=condition=Ready pod vault-0

kubectl exec vault-0 -i -- vault secrets enable pki
kubectl exec vault-0 -i -- vault secrets tune -max-lease-ttl=175200h pki
kubectl exec vault-0 -i -- vault write pki/root/generate/internal common_name=example.com key_type=ec key_bits=256 ttl=175200h
kubectl exec vault-0 -i -- vault write pki/config/urls \
    issuing_certificates="http://vault.default:8200/v1/pki/ca" \
    url_distribution_points="http://vault.default:8200/v1/pki/crl"
kubectl exec vault-0 -i -- vault write pki/roles/vault-issuer allowed_domains=cluster.local allow_subdomains=true max_ttl=48h key_type=ec key_bits=256

kubectl exec vault-0 -i -- vault auth enable kubernetes
kubectl exec vault-0 -i -- sh -c 'vault write auth/kubernetes/config \
   token_reviewer_jwt="$(cat /var/run/secrets/kubernetes.io/serviceaccount/token)" \
   kubernetes_host="https://$KUBERNETES_PORT_443_TCP_ADDR:443" \
   kubernetes_ca_cert=@/var/run/secrets/kubernetes.io/serviceaccount/ca.crt \
   issuer=https://kubernetes.default.svc.cluster.local'

kubectl exec vault-0 -i -- vault policy write vault-issuer - <<EOF
path "pki*"                 { capabilities = ["read", "list"] }
path "pki/roles/vault-issuer"   { capabilities = ["create", "update"] }
path "pki/sign/vault-issuer"    { capabilities = ["create", "update"] }
path "pki/issue/vault-issuer"   { capabilities = ["create"] }
EOF
kubectl exec vault-0 -i -- vault write auth/kubernetes/role/vault-issuer \
   bound_service_account_names=vault-issuer \
   bound_service_account_namespaces=default \
   audience=vault://default/vault-issuer \
   policies=vault-issuer \
   ttl=1m
```

Then, create an issuer and a certificate:

```sh
kubectl apply -f- <<EOF
apiVersion: v1
kind: ServiceAccount
metadata:
  name: vault-issuer
  namespace: default
---
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: vault-issuer
  namespace: default
rules:
  - apiGroups: ['']
    resources: ['serviceaccounts/token']
    resourceNames: ['vault-issuer']
    verbs: ['create']
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: vault-issuer
  namespace: default
subjects:
  - kind: ServiceAccount
    name: cert-manager
    namespace: cert-manager
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: vault-issuer
---
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: vault-issuer
  namespace: default
spec:
  vault:
    path: pki/sign/vault-issuer
    server: http://vault.default.svc.cluster.local:8200
    auth:
      kubernetes:
        role: vault-issuer
        mountPath: /v1/auth/kubernetes
        serviceAccountRef:
          name: vault-issuer # ✨
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: example-com
spec:
  secretName: example-com-tls
  issuerRef:
    name: vault-issuer
  commonName: example.cluster.local
  dnsNames:
  - example.cluster.local
  privateKey:
    algorithm: ECDSA
EOF
```

If you want to run the new e2e locally:

```sh
make -j e2e-setup-certmanager e2e-setup-vault
make e2e GINKGO_FOCUS=".*should be ready with a valid serviceAccountRef"
```

/kind feature
/area vault

```release-note
The Vault issuer can now be used with projected tokens instead of a service account Secret by using the new serviceAccountRef field.
```

maelvls commented on deploy/crds/crd-clusterissuers.yaml:
```diff
@@ -1123,7 +1123,6 @@ spec:
                           type: object
                           required:
                             - role
-                            - secretRef
```
Weird.

maelvls commented:
**Update:** last Friday, I worked on the end-to-end tests, and it now works! (cf. https://github.com/cert-manager/cert-manager/pull/5604) I also started the unit tests, but that will require another day of work. I'll plan another day of work soon.

JoshVanL commented:
By exposing the audience, we end up with a vector for an attacker to do a relay attack. This is because a user that can create Issuers but otherwise cannot impersonate a ServiceAccount, would be able to create a Vault issuer with audience set to the Kubernetes API server and point the URL to the attackers malicious server. This malicious server would then be able to impersonate the ServiceAccount. Importantly, this attack also works for other services (e.g. cloud providers etc.).

We could mitigate this by either;
1. generating the name of the audience based on the name of the Issuer (e.g. `cert-manager/ns/<namespace>/name/<issuer-name>`.
2. Always use a predefined static name or enforce a prefix to the audience which is unlikely to clash with another service (i.e. `cert-manager/vault/...`).

One could argue that this doesn't matter,  as a cluster operator needs to give both RBAC permissions to cert-manager to create tokens for specific ServiceAccounts, as well as setup the mechanism on external services to accept those tokens- both being conscious delegations of permissions. This still means however that any entity that can create Issuers that can reference that namespace can impersonate that ServiceAccount for any audience.

maelvls commented:
I agree. If one can create an Issuer in a given namespace, then they will be able to impersonate that service account (and only that one thanks to RBAC) and perform authenticated calls to the Kubernetes API. As you mentioned, that can be alleviated by making sure that the `audience` can never be set to `https://kubernetes.default.svc.cluster.local`.

The solution (1) is feasible and would allow us a stricter identity check on Vault's side. Instead of only relying on the service account's identity, cert-manager would use the audience as a way to indicate the identity of the issuer itself. On Vault's side, one would have to configure each role to match the audience calculated by cert-manager (`audiences=vault-<namespace>-<issuer-name>` for normal issuers, or `audiences=vault-<issuer-name>` for cluster-wide issuers). For example:

```sh
vault write auth/kubernetes/role/vault-issuer-1 \
   bound_service_account_names=vault-issuer-1 \
   bound_service_account_namespaces=default \
   audiences=vault-default-vault-issuer-1

vault write auth/kubernetes/role/vault-issuer-2 \
   bound_service_account_names=vault-issuer-2 \
   bound_service_account_namespaces=default \
   audiences=vault-default-vault-issuer-2
```

Even though this solution is a bit surprising because (a) the audience isn't decided by the recipient (i.e., Vault), and (b) the audience isn't fixed for a given Vault instance, I think it would not affect people because Vault's Kubernetes auth doesn't enforce a particular audience by default. We would still encourage people using the `audiences=` parameter when configuring their Vault auth roles to increase security, but it wouldn't affect people much.

The solution (2), relying on a fixed audience such as `vault`, simpler, would work well too. Again, people won't be affected (I think) because the Kubernetes auth doesn't enforce an audience by default.

japtain-cack commented:
Excitedly waiting for this to roll out. Good work everyone :)

maelvls commented:
I did some tests with Vault 1.12. Vault does not validate the audience at all, whether the parameter `audience` is passed or not. I documented the investigation in https://hackmd.io/@maelvls/vault-audience-kubernetes-auth/.

I don't know if that is a bug or not. In any case, the solutions (1) and (2) will be ineffective. I will investigate.

maelvls commented:
In [yesterday's biweekly meeting](https://docs.google.com/document/d/1Tc5t6ylY9dhXAan1OjOoldeaoys1Yh4Ir710ATfBa5U/edit#bookmark=id.vo7kmfuqb9mk), I shared the idea of going with a fixed audience string `vault`.

Ashley asked me why not go with the stricter dynamic audience of the form `cert-manager.io/my-namespace/my-issuer`. I said that this string is usually decided by the recipient (here, Vault).

In the end, I agreed to go with the dynamic audience. It won't be mandatory for users to configure it in Vault (since Vault doesn't validate the issuer), but using a dynamic audience specific to the issuer will allow for a greater security once Vault is fixed.

I will implement the dynamic audience string.

maelvls commented on pkg/apis/certmanager/v1/types_issuer.go:
```diff
@@ -269,13 +269,34 @@ type VaultKubernetesAuth struct {
 	// for authenticating with Vault. Use of 'ambient credentials' is not
 	// supported.
-	SecretRef cmmeta.SecretKeySelector `json:"secretRef"`
+	// +optional
+	SecretRef cmmeta.SecretKeySelector `json:"secretRef,omitempty"`
+	// Note: it should be a pointer because it is optional. However, for
+	// backward compatibility, we cannot change it to a pointer.
```
**question to the reviewer:** I am unsure about what to do with `SecretRef`.

Since this field becomes optional, the [api-conventions.md](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required) states that I should be using a pointer.

On the other side, changing from

```go
SecretRef cmmeta.SecretKeySelector `json:"secretRef,omitempty"`
```

to

```go
SecretRef *cmmeta.SecretKeySelector `json:"secretRef,omitempty"`
```

means that I will be breaking anyone importing this package.

I think it is fine to not use a pointer to keep backwards compatibility. I added a note to keep that in mind, since I think the inconsistency will confuse people, i.e., this:


```go
// +optional
SecretRef cmmeta.SecretKeySelector `json:"secretRef,omitempty"`

// +optional
ServiceAccountRef *ServiceAccountRef `json:"serviceAccountRef,omitempty"`
```

inteon replied:
I think that changing the field to a pointer does not break backwards compatibility towards people creating issuers from YAML manifests.

maelvls replied:
You are correct, the YAML API won't be affected.

That said, the Go API will change, and anyone relying on this `SecretRef` struct field will break.

That said:

1. The cert-manager API is not often used as a transitive dependency (if it was, it would cause much more trouble as in https://maelvls.dev/client-go-transitive-hell/).
2. The field `SecretRef` isn't usually used when importing the cert-manager API. Usually, people use the certificaterequest API for implementing an issuer.

I did a GitHub search and can't find anyone using `Issuer.spec.vault.kubernetes.secretRef`.

For these two reasons, I propose to make that breaking change anyways.

SgtCoDFish replied:
We do specifically call out `pkg/apis/` as being something that people should be safe to import in our [relevant docs](https://cert-manager.io/docs/contributing/importing/#when-you-might-import-cert-manager) on the topic.

I'm inclined to think we can't ever break anything under `pkg/apis/` without bumping the import path to cert-manager/v2 to be honest.

maelvls replied:
Agreed. I think it's fine to add a comment below SecretRef to explain why it isn't a pointer, and then to adjust the validation logic. 

And if we move to CEL one day, it will still be possible to validate that field. It will be a bit trickier, but still possible.

inteon replied:
I changed it back to a non-pointer type.

maelvls commented on internal/apis/certmanager/validation/issuer.go:
```diff
@@ -262,7 +262,80 @@ func ValidateVaultIssuerConfig(iss *certmanager.VaultIssuer, fldPath *field.Path
+	if unionCount == 0 {
+		el = append(el, field.Required(fldPath, "please supply one of: appRole, kubernetes, tokenSecretRef"))
+	}
+	if unionCount > 1 {
+		el = append(el, field.Forbidden(fldPath, "please supply one of: appRole, kubernetes, tokenSecretRef"))
+	}
```
The "at most one of" was not previously checked in [the controller code](https://github.com/cert-manager/cert-manager/blob/8444b9c101e92730fb5b454c688c03d84a8fecec/internal/vault/vault.go#L173-L207).

Adding the validation now will break anyone who previously had Issuer resources misconfigured. Users that have a misconfigured Issuer will see it in two different places:

- ~The cert-manager logs will start showing errors trying to update the Issuer spec~ Oops, my bad, we never update the Issuer spec or metadata. We only update the Issuer itself.
- The user won't be able to apply the manifest (or, say, ArgoCD will start failing).

I'll bring that up at this morning's standup.

inteon replied:
I updated the validation, PTAL @maelvls

maelvls commented:
**Update:** I struggle a lot with trying to merge the controller-side validation with the webhook-side validation. I am not confident that the validation we have added in the webhook won't break people.

I propose to remove webhook validation we have added and move it to another PR. I'll only keep the controller-side validation (the validation that is present in the `setup.go` file).

inteon commented on internal/vault/vault.go:
```diff
@@ -464,3 +513,16 @@ func (v *Vault) IsVaultInitializedAndUnsealed() error {
 	return nil
 }
+
+func (v *Vault) addVaultNamespaceToRequest(request *vault.Request) {
+	vaultIssuer := v.issuer.GetSpec().Vault
+	if vaultIssuer != nil && vaultIssuer.Namespace != "" {
+		if request.Headers != nil {
+			request.Headers.Add("X-VAULT-NAMESPACE", vaultIssuer.Namespace)
+		} else {
+			vaultReqHeaders := http.Header{}
+			vaultReqHeaders.Add("X-VAULT-NAMESPACE", vaultIssuer.Namespace)
+			request.Headers = vaultReqHeaders
+		}
+	}
+}
```
@maelvls can you remove this unused function?

maelvls replied:
Done in 5083b3e36

inteon commented on internal/vault/vault.go:
```diff
@@ -361,22 +376,56 @@ func (v *Vault) requestTokenWithAppRoleRef(client Client, appRole *v1.VaultAppRo
 
-	jwt := string(keyBytes)
+	case kubernetesAuth.ServiceAccountRef != nil:
+		aud := "vault://"
+		if v.issuer.GetNamespace() != "" {
+			aud += v.issuer.GetNamespace() + "/"
+		}
+		aud += v.issuer.GetName()
```
The audience will look like this: `vault://namespace123/issuer123` or `vault://clusterissuer123`
I think this is a valid solution, alternatives that I would have also accepted/ preferred: a static `vault` audience.

maelvls replied:
You are right, I reverted to the initial solution even though we had settled on using the hard-coded `vault`.

I added this bit of context to the PR description:

> **Alternative solution considered:** we considered using a hard-coded `vault` audience, but that did not allow to prevent the risk 1 mentioned in https://github.com/cert-manager/cert-manager/pull/5502#issuecomment-1337377810. The UX would have been better, and would correspond to what people expect from an audience (i.e., the audience should be commanded by Vault to cert-manager, not the other way around). But since this is an extra level of security (extra because Vault won't check the audience by default), it should not affect the UX much.

I hope that explains why I still went with the "generated audience".

inteon reviewed (approved):
Thank you for all the work you put into this PR
/lgtm
/approve

Please unhold when you agree that this PR can be merged.
/hold

Documentation for this feature will be added here: https://github.com/cert-manager/website/pull/1081

maelvls commented:
/unhold

maelvls commented:
**Udpate:** We just released [v1.12.0-alpha.0](https://github.com/cert-manager/cert-manager/releases/tag/v1.12.0-alpha.0) with the `serviceAccountRef` feature! I am particularly eager to hear your feedback.

More specifically, I am unsure about the decision of using an "auto-generated" audience. The consequence is that you need to create one Vault role per Issuer. You can't re-use the same Issuer or ClusterIssuer for a given Vault role. For example, I'll have to create the following role for my Issuer `vault-issuer` that is located in the namespace `sandbox`:

```bash
vault write auth/kubernetes/role/vault-issuer \
    bound_service_account_names=vault-issuer-sa \
    bound_service_account_namespaces=sandbox \
    audience="vault://sandbox/vault-issuer" \
    policies=vault-issuer
```

I cannot re-use this one Vault role with multiple Issuers or ClusterIssuers. If I have another team that needs a different ClusterIssuer, I will have to create another Vault role.

My assumption is that people will default to not using any audience (which means less security) so that they can re-use the same service account and Vault role in multiple Issuers or ClusterIssuers. For example, the following role omits `audience`, which means that Vault won't check the audience, which means the same Vault role can be used in multiple Issuers or ClusterIssuers:

```bash
vault write auth/kubernetes/role/vault-issuer \
    bound_service_account_names=team-1-sa \
    bound_service_account_namespaces=sandbox \
    policies=vault-issuer
```


If most people was to do this, it might be preferable that we switch to a hardcoded audience such as `audience=vault`.

maelvls commented:
> Without this feature, are users required to deploy both Vault and cert-Manager in the default namespace together? [...] ClusterIssuer is unable to read the service account token.

You are correct, there is no way to specify the "service account lookup namespace" used for Vault in a ClusterIssuer. The "service account lookup namespace" is conflated with the "namespace in which resources get created" and grouped under the global flag `--cluster-resource-namespace`. That's a bit unfortunate, I agree 😬

Garagoth commented:
I am having issues with audience in JWT - more details in https://github.com/cert-manager/cert-manager/issues/5942#issuecomment-1581504528
Basically kube-api complains that "vault://vault-issuer" is not a proper audience and rejects authentication.
